text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+
+  @tracked
+  get currentSeconds() {
+    return this.timer.seconds;
+  }
+
+  @tracked
+  get currentMinutes() {
+    return this.timer.minutes;
+  }
+}
+```
+
+Even though we've marked the `timer` property as tracked, the `timer.seconds`
+property is untracked, and _it_ is the field that is updated. We can solve this
+problem by using the timer library's `onTick` event handler to re-set the field,
+invalidating it:
+
+```js
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  constructor() {
+    this.timer.onTick(() => {
+      // invalidate the timer field.
+      this.timer = this.timer;
+    });
+  }
+
+  @tracked
+  get currentSeconds() {
+    return this.timer.seconds;
+  }
+
+  @tracked
+  get currentMinutes() {
+    return this.timer.minutes;
+  }
+}
+```
+
+### Interop with the Classic Programming Model
+
+Tracked properties represent a paradigm shift. They are a completely new system,
+fully independent of the classic programming model and based on modern
+Javascript features and design, and they will be the _default_ change tracking
+system in Ember Octane.
+
+However, existing apps, libraries, and addons will be using the classic
+programming model for some time, and experience tells us that these sort of
+transitions to new features take a while to settle in the community. To ease
+this process and enable gradual adoption, tracked properties will be able to
+interoperate with the most commonly used features of the classic model:
+
+- Classic classes
+- Computed properties
+- `get`/`set` and property notifications
+
+Tracked properties will _not_ interoperate with observers, which are strictly
+within the old paradigm.
+
+#### Classic Classes
+
+The `tracked` decorator function will be usable in classic classes, similar to
+`computed`:
+
+```js
+import EmberObject from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+
+const Person = EmberObject.extend({
+  firstName: tracked({ value: 'Tom' }),
+  lastName: tracked({ value: 'Dale' }),
+
+  fullName: tracked({
+    get() {
+      return `${this.firstName} ${this.lastName}`;
+    },
+  }),
+});
+```
+
+This form will _not_ be allowed on native classes, and will hard error if it is
+attempted. Additionally, default values will be defined on the _prototype_ to
+maintain consistency with the classic object model.
+
+This will allow existing libraries to transition incrementally, and add tracked
+support minimally where necessary. This also brings the _benefits_ of tracked
+to classic classes, including the ability to drop usage of `set`:
+
+```js
+// before
+let person = Person.create();
+person.set('firstName', 'Stefan');
+person.set('lastName', 'Penner');
+
+// after
+let person = Person.create();
+person.firstName = 'Stefan';
+person.lastName = 'Penner';
+```
+
+Ember's `set` function is nowhere to be seen!
+
+#### Computed Properties
+
+Computed properties will interoperate with tracked properties in both
+directions:
```

Comment:
Is there a version constraint we that goes along with this statement? As an addon author, can I still support all versions of ember I care about while implementing my package using `@tracked`?

Reply:
There will be a version constraint for addon authors unfortunately, that's good to call out. I think we can definitely polyfill/patch release 3.4 at least, I'm unsure about earlier versions of Ember. Ideally we would be able to polyfill at least 2.18.